File: src/history/tokenize.ts
/**
 * The one tokenizer the whole retrieval path shares — index and query must
 * agree or nothing matches.
 *
 * Code history is written in identifiers, so an identifier is kept WHOLE and
 * ALSO split into its parts: "retryHook" indexes as `retryhook`, `retry` and
 * `hook`, which is what makes a task phrased in prose ("the retry hook")
 * find a commit phrased in code. The whole form still scores separately, so
 * an exact identifier match stays ahead of the sum of its pieces.
 */

/**
 * Deliberately tiny. A big stopword list is a way to lose real queries —
 * "test", "error" and "run" are words this corpus is actually about.
 */
const STOPWORDS = new Set(
	(
		"the a an and or of to in on for with is are was were be been by at it its this that " +
		"from as but not we you if then so do does did has have had will would can into when " +
		"what how why all any our my me there their they"
	).split(" "),
);

/** Tokens under two characters carry no signal and blow up the posting lists. */
function keep(term: string): boolean {
	return term.length >= 2 && !STOPWORDS.has(term);
}

/**
 * "retryHook" → ["retry", "Hook"]; "retry_hook" → ["retry", "hook"];
 * "HTTPServer" → ["HTTP", "Server"] (an acronym keeps its head).
 */
function splitIdentifier(raw: string): string[] {
	return raw
		.replace(/_+/g, " ")
		.replace(/([a-z0-9])([A-Z])/g, "$1 $2")
		.replace(/([A-Z]+)([A-Z][a-z])/g, "$1 $2")
		.split(/\s+/)
		.filter((part) => part !== "");
}

/**
 * A token that is all one case with no underscore has no parts to split, and
 * that is the overwhelming majority of every corpus. Checking is one linear
 * scan; splitting it anyway costs three regex passes and a join per token —
 * ~40% of the whole index build over 2600 cards.
 */
function isPlain(raw: string): boolean {
	let upper = false;
	let lower = false;
	for (let i = 0; i < raw.length; i++) {
		const c = raw.charCodeAt(i);
		if (c === 95)
			return false; // "_"
		else if (c >= 65 && c <= 90) upper = true;
		else if (c >= 97 && c <= 122) lower = true;
		if (upper && lower) return false;
	}
	return true;
}

/**
 * Lowercased terms in occurrence order, duplicates intact — BM25 wants the
 * term frequency, so de-duplication is the caller's business.
 */
export function tokenize(text: string): string[] {
	const out: string[] = [];
	for (const raw of text.split(/[^A-Za-z0-9_]+/)) {
		if (raw === "") continue;
		const whole = raw.toLowerCase();
		if (keep(whole)) out.push(whole);
		if (isPlain(raw)) continue;
		const parts = splitIdentifier(raw);
		if (parts.length < 2) continue;
		for (const part of parts) {
			const term = part.toLowerCase();
			if (keep(term)) out.push(term);
		}
	}
	return out;
}

/** Query terms, de-duplicated: repeating a word must not double its weight. */
export function queryTerms(text: string): string[] {
	return [...new Set(tokenize(text))];
}

/**
 * Anchored character trigrams ("hook" → ^ho, hoo, ook, ok$) — the candidate
 * generator for fuzzy matching. Anchoring matters: it keeps a typo at the
 * head or tail of a word visible instead of averaging it away.
 */
export function trigrams(term: string): string[] {
	const padded = `^${term.toLowerCase()}$`;
	const out: string[] = [];
	for (let i = 0; i + 3 <= padded.length; i++) out.push(padded.slice(i, i + 3));
	return out;
}

File: crates/flusk-core/src/index/tokenize.rs
//! Port of src/features/history/tokenize.ts. Index and query must agree with
//! the TS reference token for token — the golden differential test holds both
//! to the same output.

const STOPWORDS: &[&str] = &[
    "the", "a", "an", "and", "or", "of", "to", "in", "on", "for", "with", "is", "are", "was",
    "were", "be", "been", "by", "at", "it", "its", "this", "that", "from", "as", "but", "not",
    "we", "you", "if", "then", "so", "do", "does", "did", "has", "have", "had", "will", "would",
    "can", "into", "when", "what", "how", "why", "all", "any", "our", "my", "me", "there",
    "their", "they",
];

fn keep(term: &str) -> bool {
    term.len() >= 2 && !STOPWORDS.contains(&term)
}

/// "retryHook" → ["retry","Hook"]; "HTTPServer" → ["HTTP","Server"] — the
/// acronym keeps its head, exactly as the TS regex pair splits it.
fn split_identifier(raw: &str) -> Vec<String> {
    let mut spaced = String::with_capacity(raw.len() + 8);
    let bytes: Vec<char> = raw.chars().collect();
    for (i, &c) in bytes.iter().enumerate() {
        if c == '_' {
            spaced.push(' ');
            continue;
        }
        // ([a-z0-9])([A-Z]) boundary
        if i > 0 && c.is_ascii_uppercase() {
            let prev = bytes[i - 1];
            if prev.is_ascii_lowercase() || prev.is_ascii_digit() {
                spaced.push(' ');
            } else if prev.is_ascii_uppercase() {
                // ([A-Z]+)([A-Z][a-z]) boundary: an upper before an Upper-lower pair
                if bytes.get(i + 1).is_some_and(|n| n.is_ascii_lowercase()) {
                    spaced.push(' ');
                }
            }
        }
        spaced.push(c);
    }
    spaced.split_whitespace().map(str::to_string).collect()
}

/// All one case, no underscore: nothing to split — the fast path that is ~40%
/// of a full index build in the reference implementation.
fn is_plain(raw: &str) -> bool {
    let mut upper = false;
    let mut lower = false;
    for c in raw.chars() {
        if c == '_' {
            return false;
        } else if c.is_ascii_uppercase() {
            upper = true;
        } else if c.is_ascii_lowercase() {
            lower = true;
        }
        if upper && lower {
            return false;
        }
    }
    true
}

/// Lowercased terms in occurrence order, duplicates intact.
pub fn tokenize(text: &str) -> Vec<String> {
    let mut out = Vec::new();
    for raw in text.split(|c: char| !(c.is_ascii_alphanumeric() || c == '_')) {
        if raw.is_empty() {
            continue;
        }
        let whole = raw.to_ascii_lowercase();
        if keep(&whole) {
            out.push(whole);
        }
        if is_plain(raw) {
            continue;
        }
        let parts = split_identifier(raw);
        if parts.len() < 2 {
            continue;
        }
        for part in parts {
            let term = part.to_ascii_lowercase();
            if keep(&term) {
                out.push(term);
            }
        }
    }
    out
}

/// Query terms, de-duplicated in first-seen order.
pub fn query_terms(text: &str) -> Vec<String> {
    let mut seen = std::collections::HashSet::new();
    tokenize(text)
        .into_iter()
        .filter(|t| seen.insert(t.clone()))
        .collect()
}

/// Anchored character trigrams ("hook" → ^ho, hoo, ook, ok$).
pub fn trigrams(term: &str) -> Vec<String> {
    let padded: Vec<char> = format!("^{}$", term.to_ascii_lowercase()).chars().collect();
    let mut out = Vec::new();
    let mut i = 0;
    while i + 3 <= padded.len() {
        out.push(padded[i..i + 3].iter().collect());
        i += 1;
    }
    out
}
